validate mail addresses during event import and handle RecipientsNotFoundError for calendar events

vaf-hub opened the pull request:
closes  #2975

I decided to ignore incorrectly formatted attendee and organizer mail addresses during calendar imports because we already ignore certain attendees and organizers when importing calendars. The alternative to allow invalid addresses would mean that we would have to check for validity of each address in many places and omission to do so might lead to more bugs.

Also, this PR fixes RecipientsNotFoundError handling of non-existing tutanota addresses in calendar events. It is now possible to delete such events or remove the incorrect addresses from the list of attendees. However, sending out updates for such events will still show an error dialog that mentions the invalid attendee addresses. So in this case the user has to manually cleanup by deleting the respective addresses (or replacing them with correct ones) before sending the update.

johnbotris commented on src/calendar/CalendarUpdateDistributor.js:
```diff
@@ -72,6 +73,21 @@ export class CalendarMailDistributor implements CalendarUpdateDistributor {
 			event,
 			sender
+		}).catch(RecipientsNotFoundError, e => {
+			// we want to delete the event even if the recipient is not an existing tutanota address
```
is this comment correct? i'm not sure it matches what im seeing

vaf-hub replied:
I think so. This refers to the following case: the user wants to delete a calendar event but there are invalid (attendee) addresses in it. usually we send out a mail that the event was cancelled to all attendees, but in this case we can't and receive an error. so we can't delete the event. therefore, i decided, it was best to ignore invalid recipients and only send the notifications to those recipients that we can resolve the mail address for. i thought, it would be to annoying to bother the user to manually remove or correct these mail addresses before being able to delete an event. we can discuss it, though.
anyway, prior to this there was already another similar case: external attendees without a password of a confidential event also will not receive a notification when we delete the event.

vaf-hub replied:
I will drop the last commit

johnbotris commented on src/calendar/CalendarUpdateDistributor.js:
```diff
@@ -72,6 +73,21 @@ export class CalendarMailDistributor implements CalendarUpdateDistributor {
+		}).catch(RecipientsNotFoundError, e => {
+			// we want to delete the event even if the recipient is not an existing tutanota address
+			// and just exclude them from sending out updates but leave the event untouched for other recipients
+			const invalidRecipients = e.message.split("\n")
```
can we be sure that the message will always have the expected format?

johnbotris replied:
I would imagine just something like this:

```
export class RecipientsNotFoundError extends TutanotaError {

	+recipients: Array<string>

	constructor(recipients: Array<string>) {
		super("RecipientsNotFoundError",  recipients.join("\n"))
	}
}
```
 

> I am not sure we want to create this one error in a different way than all others.

I think it would be fine to do, we just haven't done it yet because we haven't needed to

vaf-hub replied:
Yeah, I just looked into it a bit more and I am still not convinced we want to make such a change. The reason is that currently we pass errors between the worker and main thread in messages and we would have to handle this error differently from all others then, because we cannot simply pass the message as it is the only one where we do not pass a string to the constructor. Take a look at: `objToError()` which is what we use to recreate the error after we receive such a message from the worker.

vaf-hub replied:
So, I pushed another commit, but it does not work of course until we handle the beforementioned problem

johnbotris replied:
Ah I see, yeah that complicates things. Better to leave it how it is then, making it work nicely would probably be a bigger change than it's worth

johnbotris commented:
Tests are failing: 

```
CalendarEventViewModel > delete event > own event with non-existing internal attendees in own calendar:
––––––
makeMailModel is not defined
All 2599 assertions passed (old style total: 2808). Bailed out 1 time  
    at eval (nollup-int:///client/calendar/CalendarEventViewModelTest.js:384:25)
    at processTicksAndRejections (internal/process/task_queues.js:75:11)
```

vaf-hub commented:
> Tests are failing:
> 
> ```
> CalendarEventViewModel > delete event > own event with non-existing internal attendees in own calendar:
> ––––––
> makeMailModel is not defined
> All 2599 assertions passed (old style total: 2808). Bailed out 1 time  
>     at eval (nollup-int:///client/calendar/CalendarEventViewModelTest.js:384:25)
>     at processTicksAndRejections (internal/process/task_queues.js:75:11)
> ```

The function was removed in a different commit merged to master recently.
I fixed the error by removing the one test case I had added in this branch but that did not make too much sense because of the mocks we had, anyway.
